### DISCUSSION TITLE
Encryption support

### DISCUSSION THREAD

**Discussion by mjablecnik** [isar/isar#176] [Ideas]
Hello, do you plan also support encryption similar like Hive or Sembast?

**Comment by simc** [comment]
Unfortunately not. Entire collections will not support encryption but I'm thinking about implementing property-based encryption. Would you be interested in that?

Could you please share your use-case to allow me to understand what you need exactly?

**Reply by mjablecnik** [reply]
I work for one package delivery company and I create application for couriers which deliver packages to our clients.

Application have to work online and offline because somewhere can be some bad connection. Every morning I will save information about packages and clients for current day and update it during communication with server. When app will become offline so it must work only with local database and after connection restore sync changes with server.

In local database I will have a lot of sensitive information about clients (name, surname, address, phone, etc.) and delivered packages. Maybe also some documents. This information must be encrypted because phone device can be stolen or forgotten somewhere and data must be saved somehow.

**Reply by simc** [reply]
Thanks, that helps a lot. Would it be enough for you to encrypt certain properties?

There are more important tasks currently but I'll definitely keep it in mind and try to come up with a solution.

**Reply by mjablecnik** [reply]
I don't know. Now I am doing only some research which database/package I can use (hive, sembast, sqflite_sqlcipher or something else) for my purposes. I prefer some simple solution and if I can encrypt all DB so I will use it than encrypt all every single properties..
I don't know how big problem it is for you to encrypt whole collection but I will start work on my application probably from February or March so if you will add this functionality in some next 2 months it would be good and I will try it. Otherwise I will use probably some other solution/package.

Btw.: I think this functionality can be useful also for other people than only for me in the future.. ;)

**Reply by MarcelGarus** [reply]
Also note that nowadays, both Android and iOS also implement filesystem-level encryption by default. This means every file is encrypted by the operating system itself. You can't simply extract data from a stolen device.

**Reply by simc** [reply]
Completely agree. It's even hard to get internal app data from an app if the device is **unlocked** 😅

**Reply by mjablecnik** [reply]
@MarcelGarus Thank you very much for this information. So when I will use `path_provider`:
```
final dir = await getApplicationSupportDirectory(); 
```
So all my data will be saved and encrypted there by default, is it true?

I also studied this page: https://source.android.com/security/encryption/full-disk
And I see there a note: 
![image](https://user-images.githubusercontent.com/2686906/148920582-152b19b3-9981-4a21-a5cc-d29be4c50b2a.png)
This looks like in Android from version 10 must everything be encrypted, is it true? Does Flutter do this by default so I cannot worry about any encryption and security issues?

Thank you very much

**Reply by MarcelGarus** [reply]
Yes! Encryption happens automatically! You don't need to worry about it.

Here's a little bit of history that helps understanding the screenshot:

- Before Android 7, Android didn't support encryption out-of-the-box. Apps had to take care of encrypting the content.
- Since Android 7, you could optionally encrypt your disk, which means no functionality was available until you entered your PIN. The disadvantage was that you could basically do nothing without entering the PIN. So, for example, if the phone updated overnight, the alarm would not go off because the alarm couldn't read from the disk. Similarly, receiving phone calls wouldn't work until you unlocked your phone at least once since booting it. That's why this feature was opt-in.
- Since Android 10, file-based encryption is used instead: The operating system itself decides which files to encrypt with which key. Some special files from the system, the alarm clock etc. are only encrypted with a key that is stored on disk (on some devices in a secure enclave, so it's harder to physically extract the key). But by default, files are encrypted using a key derived from the user's PIN. Because this doesn't limit the functionality of the device, this is enabled on all devices.

So for all apps (with only a handful of exceptions), on Android 10, the encryption of the data happens automatically. If you just save something in the application support directory, it will be encrypted by the operating system.

That being said, I'm not a lawyer, so this is not any legally binding advice.

**Reply by mjablecnik** [reply]
Ok so for Android 7,8 and 9 user must encrypt their disk himself. From Android 10 it is automatically.
But when user will not encrypt their disk (when it is optional) so it will be not secure still, is it true?

**Reply by MarcelGarus** [reply]
Yes, that's true. Data on Android 10+ is encrypted for sure. For Android 7, 8, and 9, if the device gets stolen and disk encryption is not turned on (which is on most devices), then the data can be extracted.

**Reply by mjablecnik** [reply]
So I will need to care about encryption because I have to support also versions older then Android 10..

**Reply by florian-lefebvre** [reply]
I'm also interested in this, Hive encryption was great and I would like to encrypt the data of my users with Isar. If you plan to implement property-based encryption, it could be enough but I guess total encryption is better for privacy. Still, I'm not the maintainer and do not have a global vision of Isar so can't say what's best or possible but just giving my opinion.

Also, having an issue dedicated to Encryption to track its progress would be great.

**Reply by tbleiker** [reply]
I have to store sensitive data in a cross-platform app. Therefore, I need an encrypted DB at least for the desktop app (since I cannot guarantee that the db file is stored on an encrypted filesystem). 

Isar looks promising, I hope that full-encryption will be considered ;-)

**Reply by khainke** [reply]
We also want to store sensitive data for an app, that supports mac, windows, android and ios. The data might include personal information, that is subject to data privacy laws and that a user most likely wants to store encrypted, but not all users might encrypt their entire filesystem.

**Reply by MuhammedKpln** [reply]
Any news?

**Reply by craftedcodes** [reply]
I would love to encrypt my Isar db. I want to store sensitive data and currently, I have to decide between Hive, Floor and Isar and the encryption part is the main reason on which my choice is depending.

**Reply by simc** [reply]
Isar v4 supports encryption

**Comment by jacob-emery** [comment]
How am I supposed to securely store access token and secret in Android? Would you use Hive with FlutterSecureStorage or only FlutterSecureStorage for that? Such as in https://github.com/hivedb/docs/blob/master/advanced/encrypted_box.md ? Or do you think that, because of file-based encryption, using Isar would be okay (Android 10+)?

**Reply by yevyevyev** [reply]
tokens and secrets are stored in local keychains of the OS, use whatever library wraps the native API's to access keychain.

**Comment by skavinvarnan** [comment]
As isar is also supported for web,  it would be great to have encryption for the whole database. Since its pretty easy to get the DB file from the browser

**Reply by oliverbytes** [reply]
> As isar is also supported for web, it would be great to have encryption for the whole database. Since its pretty easy to get the DB file from the browser

I second this. The db file on Desktop would be easily hacked as well. Encryption support is vital

**Reply by simc** [reply]
Unfortunately IndexedDB has no support for encryption at all. It is the recommended database backend on Web and for example used by WhatsApp web

**Reply by oliverbytes** [reply]
> Unfortunately IndexedDB has no support for encryption at all. It is the recommended database backend on Web and for example used by WhatsApp web

Thanks Simon! Encryption support except Web would at least suffice some use cases. Hopeful for that day to come. Thanks for your super great work!!!!

**Comment by soumyadeep-ghosh** [comment]
Any Updates on the Encryption that can be used in 7,8,9 ? Looking forward to create a DB with Sensitive Information, Property Based Encryption also works.

**Reply by elhe26** [reply]
Linking this discussion to #292 .
